search: Refactor and polish pills implementation.

ryanreh99 opened the pull request:
This is a WIP PR for the polishing work on search pills and a continuation of #14930.
( Will create a new PR when we want to migrate to the search_pills_enabled version. )

- [ ] If pills overflow, they break the 2 search icons.
- [ ] If the cursor is at the start of the input block and text exists, then it is not possible to delete the previous pill. This is an input pills specific problem and not a search pills one.

timabbott commented:
The first two commits lgtm; not merging since you called this WIP, and that's ambiguous for which commits you're happy with.  

For the third commit, what's the reasoning for why that change won't cause problems with other places we use pills?  I haven't thought through it carefully.

Also @shubham-padia if you have a bit of time, it'd be great if you can at least do a bit of QA on this branch.

timabbott commented:
Are any of these commits ready for me to review and merge?  I generally like to do so to avoid branches getting too large and complicated without review, but am not going to do so without explicit feedback that you think a set of commits is ready, given the mention that this WIP.

ryanreh99 commented:
@timabbott the first 3 commits are ready for review.

> For the third commit, what's the reasoning for why that change won't cause problems with other places we use pills?

Since we are clearing the input box after creating the pill, I don't think it matters if the comma is typed or not.
I think this is a bugfix because on not using the type-ahead for the private message compose box, we get the same error i.e. pills gets created, and input is cleared then comma is added.

timabbott commented:
I merged the first 3 commits.  @showell would you be up for helping review and integrate the next couple unmerged commits?  I think they're updating the pills-based search variant to match the logic you recently rewrote for the non-legacy case, and you're perfectly situated to understand that refactoring effort.

ryanreh99 commented:
I think the only difference between the pills and legacy version are these 2:
• We want to generate suggestions (based on previous query) even if no input is entered (but pills exist). The `else if` block which pushes an empty last term is responsible for this.
• The `get_operator_subset_suggestions` function where as mentioned in the comment, it would be easier to just delete the pills.

____________________________________

The gap between the pills and input is a recent regression being tracked [here](https://chat.zulip.org/#narrow/stream/101-design/topic/navbar.20redesign/near/895413) and [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/nightmode.20margin.20regression).

ryanreh99 commented:
@showell could you take a look.
( I also updated the task list to include the bugs I noticed. )

showell commented:
@ryanreh99 Thanks for working on this!  I will be so glad when we can get rid of all the legacy code.

Can you help me review this by making your commit messages a little more detailed?

Here is the first one I'm reviewing:

~~~
    search suggestions: Refactor get_search_result code path.

    This is a prep commit to make it simplier for when we combine
    the `get_suggestions` and `get_suggestions_legacy` functions.
~~~

There's a typo with "simplier", but I'm not concerned about that, of course.

I would try to just explain the changes in a bit more detail.  Here is my best attempt:

~~~
<what you already said>

This codepath mostly changes things in the **new** search code, which is not yet
turned on.  That is mostly evident from the fact that we don't need to make any
changes to `frontend_tests/node_tests/search_suggestion_legacy.js`.

The old code now calls `get_default_suggestion` instead of `get_default_suggestion_legacy`,
but that function still uses the **old** implementation.

Then most of the substantive changes here happen in `get_search_result`, which is
part of the new codepath.

A lot of the diff inside `get_search_result` is just renaming `query_operators` to
`search_operators`.  (EDITOR NOTE: is there some reason this couldn't have happened
in a separate commit?)

We also make changes in the `get_search_result_legacy` function that is still what
actually runs on master.  Here, we are mostly renaming `operators` to `search_operators`.
(EDITOR NOTE: is there some reason that couldn't have happened in a separate commit?)
~~~

Hopefully you get the idea.  I actually do wonder why this commit couldn't have been broken down into a few more atomic steps.

Finally, I would mention the one-line change to the new tests.  Do you know why the old test stopped passing.  (It's good to explain in a commit that you simply punted on it, because you didn't believe the old test made sense, but if you can explain that more, it would be helpful.  I was able to confirm locally that the test change there was in fact necessary, so it wasn't a spurious change.)

ryanreh99 commented:
@showell @timabbott This is ready for another review.

I updated the commit messages.
Now the only bugs remaining are (as mentioned [above](https://github.com/zulip/zulip/pull/15172#issue-426397615)): 

• The searchbox experience can be improved if the user uses the mouse instead of keyboard shortcuts.
Points 1 and 2: It is hard continuing typing when clicking on a typeahead suggestion or deleting a pill due to the "focusout" event.

• This is only when typing multiple queries (Point 3): When selecting a value from the typeahead suggestions, the searchbox gets re-populated (automated selection) with the comma-separated queries (which is how it should be stored due to [this reason](https://chat.zulip.org/#narrow/stream/6-frontend/topic/search.20pills.20.2314930/near/880718)) but we should populate with the space-separated queries.

![ci](https://user-images.githubusercontent.com/40122794/84915287-382c2e00-b0da-11ea-86d3-c61727668cf2.png)

ryanreh99 commented:
last push was just fixing typos in the commit messages.
( Posting the ci pic on green... )

timabbott commented:
I merged most of this as the series ending with a7dc0ed816a50122d4933ccd493d4c6ec364cd8b

> EDIT: Commit "typeahead: Prevent lookup on backspace keyup event." breaks for the other places pills are used.

I was going to make the same comment :)

timabbott commented:
I squashed the following change into the big deduplication commit, which makes it a lot more readable in my view:
```
diff --git a/static/js/search_suggestion.js b/static/js/search_suggestion.js
index 4547e7a214..3caca26618 100644
--- a/static/js/search_suggestion.js
+++ b/static/js/search_suggestion.js
@@ -609,7 +609,7 @@ function make_attacher(base) {
 
 exports.get_search_result = function (base_query, query) {
     let suggestion;
-    let operators;
+    let all_operators;
 
     // search_operators correspond to the operators for the query in the input.
     // For search_pills_enabled, this includes just editable query where search pills
@@ -617,7 +617,7 @@ exports.get_search_result = function (base_query, query) {
     // And for this disabled case, this includes the entire query entered in the searchbox.
     // operators correspond to the operators for the entire query entered in the searchbox.
     if (page_params.search_pills_enabled) {
-        operators = Filter.parse((base_query + " " + query).trim());
+        all_operators = Filter.parse((base_query + " " + query).trim());
     }
     const search_operators = Filter.parse(query);
     let last = {operator: '', operand: '', negated: false};
@@ -629,7 +629,7 @@ exports.get_search_result = function (base_query, query) {
         // calculated from the created search pills.
         // Else search results are returned as if the user is still
         // typing the non-editable last search pill.
-        operators.push(last);
+        all_operators.push(last);
         search_operators.push(last);
     }
 
@@ -653,8 +653,8 @@ exports.get_search_result = function (base_query, query) {
                 negated: person_op.negated,
             };
             if (page_params.search_pills_enabled) {
-                operators[operators.length - 2] = last;
-                operators.splice(-1, 1);
+                all_operators[all_operators.length - 2] = last;
+                all_operators.splice(-1, 1);
             }
             search_operators[search_operators_len - 2] = last;
             search_operators.splice(-1, 1);
@@ -697,7 +697,9 @@ exports.get_search_result = function (base_query, query) {
         get_has_filter_suggestions,
     ];
 
-    const all_operators = page_params.search_pills_enabled ? operators : search_operators;
+    if (!page_params.search_pills_enabled) {
+        all_operators = search_operators;
+    }
     const base_operators = all_operators.slice(0, -1);
     const max_items = exports.max_num_of_search_results;
 
```

timabbott commented:
Merged that as 69f9c72edc5fd8c87bb6895517784c4a07e15e3f, thanks @ryanreh99!

Let me know when you have a solution to the bootstrap issue.  (As a general rule, we avoid patching bootstrap, because every time we do that adds more work to unforking it in the future).  

If we can, I'd like to test turning this on at chat.zulip.org to get more QA on it tomorrow.

ryanreh99 commented:
> As a general rule, we avoid patching bootstrap, because every time we do that adds more work to unforking it in the future)

Ah yes I understand, updated the commit to include these changes inside an `if` block.
Also I was wondering maybe we should document the `helpOnEmptyStrings` option on top of this `typeahead.js` file as this is also a zulip patch from commit [5574efd](https://github.com/zulip/zulip/commit/d3293175021c03e21e11599bad5faa47c0a67b15).
And we could also mention about the `delete` key function in the `Search Operators` info overlay or in a new "Search" section under the `Keyboard Shortcuts` overlay.

timabbott commented on static/js/search.js:
```diff
@@ -180,6 +180,14 @@ exports.initialize = function () {
+            // lose focus when a pill inside it gets selected
+            // or deleted by a click.
+            return;
+        }
```
Does this selector need to be more specific?  I worry about clicking on a pill in the compose box, or something, being misinterpreted

timabbott commented:
> Also I was wondering maybe we should document the helpOnEmptyStrings option on top of this typeahead.js file as this is also a zulip patch from commit 5574efd.

Yes, definitely, the better documentation we have for typeahead changes the more able to manage them in the future we will be.

ryanreh99 commented:
@timabbott could you take a look.
I think the search pills version is usable now.
( I updated the task list at the top of the issue. )

![ci](https://user-images.githubusercontent.com/40122794/85163062-e91ffd80-b27f-11ea-8a7c-626902f220f9.png)

timabbott commented:
@YashRE42 can you do some thorough testing on the legacy search behavior changes in this PR? 

Ideally we'd be able to merge this and then enable search pills on chat.zulip.org.

@ryanreh99 I don't understand why we would expect the typeahead.js changes to not cause changes in other typeaheads, e.g. in the compose box.  Can you talk through your reasoning on that front?

ryanreh99 commented:
> @ryanreh99 I don't understand why we would expect the typeahead.js changes to not cause changes in other typeaheads, e.g. in the compose box. Can you talk through your reasoning on that front?

Oops while writing the reasoning, realised this was mostly copied code.
Updated the commit. (I removed the first parameter as it was unused. )

timabbott commented:
OK, in this revised version, I think I'm happy with the changes.  Merged, thanks @ryanreh99!

I'm deploying on chat.zulip.org now with SEARCH_PILLS_ENABLED=True so we can get some testing.
